nixos/doc/manual/development/freeform-modules.xml

```diff
@@ -0,0 +1,69 @@
+<section xmlns="http://docbook.org/ns/docbook"
+        xmlns:xlink="http://www.w3.org/1999/xlink"
+        xmlns:xi="http://www.w3.org/2001/XInclude"
+        version="5.0"
+        xml:id="sec-freeform-modules">
+ <title>Freeform modules</title>
+ <para>
+  Freeform modules allow you to define values for option paths that have not been declared explicitly. This can be used to add attribute-specific types to <literal>attrsOf</literal> options.
+ </para>
+ <para>
+  This feature can be enabled by using the option <literal>_module.freeformType</literal> to define a freeform type. By doing this, all assigments without an associated option will get merged using the freeform type and combined into the resulting <literal>config</literal> set. Since this feature influences the whole module evaluation, it is mostly recommended for submodule options only.
+ </para>
+ <example xml:id="ex-freeform-module">
+  <title>Freeform submodule</title>
+  <para>
+   The following shows a submodule assigning a freeform type that allows arbitrary <literal>str</literal> attribute values, but also declares an option for the <literal>port</literal> attribute to have it type-checked and assign a default value.
+  </para>
+ <programlisting>
+{ lib, config, ... }: {
+
+  options.settings = lib.mkOption {
+    default = {};
+    type = lib.types.submodule {
+
+      config._module.freeformType = with lib.types; attrsOf str;
```

Comment:
```suggestion
      freeformType = with lib.types; attrsOf str;
```

Can we add syntax sugar like this? Besides being a bit cryptic, it's kind of in the wrong part of the module syntax.

(This suggestion applies to other mentions of `_module.freeformType` as well of course)

Reply:
I haven't considered this yet. I don't see an immediate problem with it, other than it making the implementation a bit more involved. I'll give it some thought